algorithms/maths/rsa.py

```diff
@@ -0,0 +1,93 @@
+"""
+RSA encryption algorithm
+a method for encrypting a number that uses seperate encryption and decryption keys
+this file only implements the key generation algorithm
+
+there are three important numbers in RSA called n, e, and d
+e is called the encryption exponent
+d is called the decryption exponent
+n is called the modulus
+
+these three numbers satisfy
+((x ** e) ** d) % n == x % n
+
+to use this system for encryption, n and e are made publicly available, and d is kept secret
+a number x can be encrypted by computing (x ** e) % n
+the original number can then be recovered by computing (E ** d) % n, where E is
+the encrypted number
+
+fortunately, python provides a three argument version of pow() that can compute powers modulo
+a number very quickly:
+(a ** b) % c == pow(a,b,c)
+"""
+
+# sample usage:
+# n,e,d = generate_key(16)
+# data = 20
+# encrypted = pow(data,e,n)
+# decrypted = pow(encrypted,d,n)
+# assert decrypted == data
+
+import random
+
+
+def generate_key(k, seed=None):
+    """
+    the RSA key generating algorithm
+    k is the number of bits in n
+    """
+
+    def modinv(a, m):
+        """calculate the inverse of a mod m
+        that is, find b such that (a * b) % m == 1"""
+        b = 1
+        while not (a * b) % m == 1:
+            b += 1
+        return b
+
+    def gen_prime(k, seed=None):
+        """generate a prime with k bits"""
+
+        def is_prime(num):
+            if num == 2:
+                return True
+            for i in range(2, int(num ** 0.5) + 1):
+                if num % i == 0:
+                    return False
+            return True
+
```

Comment:
Consider using a more efficient primality test like Miller-Rabin for generating large primes, especially for longer key sizes.